File: day34/inorderTraversal.java
// 给定一个二叉树，返回它的中序遍历。

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 非递归
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();

        while (!stack.isEmpty() || root != null) {
            if (root != null) {
                stack.push(root);
                root = root.left;
            } else {
                TreeNode cur = stack.pop();
                res.add(cur.val);
                root = cur.right;
            }
        }
        return res;
    }
}

// 递归
// class Solution {
//     public List<Integer> inorderTraversal(TreeNode root) {
//         List<Integer> res = new ArrayList<>();
//         dfs(root, res);
//         return res;
//     }

//     private void dfs(TreeNode root, List<Integer> res) {
//         if (root == null) return;
//         dfs(root.left, res);
//         res.add(root.val);
//         dfs(root.right, res);
//         return;
//     }
// }

File: day34/inorderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        stack = []
        
        while stack or root:
            if root:
                stack.append(root)
                root = root.left
            else:
                cur = stack.pop()
                res.append(cur.val)
                root = cur.right
        return res

# class Solution:
#     def inorderTraversal(self, root: TreeNode) -> List[int]:
#         if not root: return []
#         stack = [(True, root)]
#         res = []

#         while stack:
#             flag, cur = stack.pop()
#             if flag:
#                 if cur.right: stack.append((True, cur.right))
#                 stack.append((False, cur))
#                 if cur.left: stack.append((True, cur.left))
#             else:
#                 res.append(cur.val)
#         return res

# class Solution:
#     def inorderTraversal(self, root: TreeNode) -> List[int]:
#         res = []
#         def dfs(root, res):
#             if not root: return
#             dfs(root.left, res)
#             res.append(root.val)
#             dfs(root.right, res)
#             return
#         dfs(root, res)
#         return res
